diff --git a/src/fread.c b/src/fread.c
--- a/src/fread.c
+++ b/src/fread.c
@@ -216,6 +216,7 @@ char *fread_string (FILE *fp, char *buf, size_t size) {
     char *plast;
     char c;
 
     plast = buf;
     c = fread_letter (fp);
     if ((*plast++ = c) == '~')
         return str_empty;
@@ -228,25 +229,30 @@ char *fread_string (FILE *fp, char *buf, size_t size) {
          * it was too dirty for portability.
          *   -- Furey */
 
-        switch (*plast = getc (fp)) {
-            case EOF:
-                /* temp fix */
-                bug ("fread_string: EOF", 0);
-                return NULL;
-
-            case '\n':
-                plast++;
-                *plast++ = '\r';
-                break;
-
-            case '\r':
-                break;
-
-            case '~':
-                *plast = '\0';
-                return buf;
-
-            default:
-                plast++;
-                break;
-        }
+        int ich = getc(fp);
+
+        switch (ich) {
+        case EOF:
+            /* temp fix */
+            bug("fread_string: EOF", 0);
+            return NULL;
+
+        case '\n':
+            *plast++ = '\n';
+            *plast++ = '\r';
+            break;
+
+        case '\r':
+            break;
+
+        case '~':
+            *plast = '\0';
+            return buf;
+
+        default:
+            *plast++ = (char)ich;
+            break;
+        }
     }
 }
@@ -271,6 +277,7 @@ char *fread_string_eol (FILE *fp, char *buf, size_t size) {
     static bool char_special[256 - EOF];
     char *plast;
 
     if (char_special[EOF - EOF] != TRUE) {
         char_special[EOF - EOF]  = TRUE;
         char_special['\n' - EOF] = TRUE;
         char_special['\r' - EOF] = TRUE;
@@ -284,25 +291,32 @@ char *fread_string_eol (FILE *fp, char *buf, size_t size) {
 
     while (1) {
-        if (!char_special[(*plast++ = getc (fp)) - EOF])
-            continue;
-
-        switch (plast[-1]) {
-            case EOF:
-                bug ("fread_string_eol: EOF", 0);
-                exit (1);
-                break;
-
-            case '\n':
-            case '\r':
-                plast--;
-                *plast = '\0';
-                return buf;
-
-            default:
-                break;
-        }
+        int ich = getc(fp);
+
+        if (ich == EOF) {
+            bug("fread_string_eol: EOF", 0);
+            exit(1);
+        }
+
+        *plast++ = (char)ich;
+
+        if (!char_special[(unsigned char)ich - EOF])
+            continue;
+
+        switch (ich) {
+        case '\n':
+        case '\r':
+            plast--;
+            *plast = '\0';
+            return buf;
+
+        default:
+            break;
+        }
     }
 }
